collect_app/src/main/java/org/odk/collect/android/map/MapboxMapFragment.java

```diff
@@ -0,0 +1,662 @@
+
+import android.content.SharedPreferences;
+import android.graphics.drawable.Drawable;
+import android.location.Location;
+import android.os.Handler;
+import android.preference.PreferenceManager;
+import android.view.Gravity;
+
+import com.mapbox.android.core.location.LocationEngine;
+import com.mapbox.android.core.location.LocationEngineCallback;
+import com.mapbox.android.core.location.LocationEngineProvider;
+import com.mapbox.android.core.location.LocationEngineRequest;
+import com.mapbox.android.core.location.LocationEngineResult;
+import com.mapbox.mapboxsdk.Mapbox;
+import com.mapbox.mapboxsdk.camera.CameraUpdate;
+import com.mapbox.mapboxsdk.camera.CameraUpdateFactory;
+import com.mapbox.mapboxsdk.geometry.LatLng;
+import com.mapbox.mapboxsdk.geometry.LatLngBounds;
+import com.mapbox.mapboxsdk.location.LocationComponent;
+import com.mapbox.mapboxsdk.location.LocationComponentActivationOptions;
+import com.mapbox.mapboxsdk.location.LocationComponentOptions;
+import com.mapbox.mapboxsdk.location.modes.CameraMode;
+import com.mapbox.mapboxsdk.location.modes.RenderMode;
+import com.mapbox.mapboxsdk.maps.MapboxMap;
+import com.mapbox.mapboxsdk.maps.OnMapReadyCallback;
+import com.mapbox.mapboxsdk.maps.Style;
+import com.mapbox.mapboxsdk.plugins.annotation.Line;
+import com.mapbox.mapboxsdk.plugins.annotation.LineManager;
+import com.mapbox.mapboxsdk.plugins.annotation.LineOptions;
+import com.mapbox.mapboxsdk.plugins.annotation.OnSymbolDragListener;
+import com.mapbox.mapboxsdk.plugins.annotation.Symbol;
+import com.mapbox.mapboxsdk.plugins.annotation.SymbolManager;
+import com.mapbox.mapboxsdk.plugins.annotation.SymbolOptions;
+import com.mapbox.mapboxsdk.style.layers.RasterLayer;
+import com.mapbox.mapboxsdk.style.layers.TransitionOptions;
+import com.mapbox.mapboxsdk.style.sources.RasterSource;
+import com.mapbox.mapboxsdk.style.sources.TileSet;
+import com.mapbox.mapboxsdk.utils.ColorUtils;
+
+import org.odk.collect.android.BuildConfig;
+import org.odk.collect.android.R;
+import org.odk.collect.android.application.Collect;
+import org.odk.collect.android.mapboxsdk.MapFragment;
+import org.odk.collect.android.preferences.GeneralKeys;
+
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+
+import androidx.annotation.NonNull;
+import androidx.annotation.Nullable;
+import androidx.annotation.VisibleForTesting;
+import androidx.core.content.ContextCompat;
+import androidx.fragment.app.Fragment;
+import androidx.fragment.app.FragmentActivity;
+import edu.umd.cs.findbugs.annotations.SuppressFBWarnings;
+
+import static android.os.Looper.getMainLooper;
+
+public class MapboxMapFragment extends MapFragment implements org.odk.collect.android.map.MapFragment,
+    OnMapReadyCallback, MapboxMap.OnMapClickListener, MapboxMap.OnMapLongClickListener,
+    LocationEngineCallback<LocationEngineResult> {
+
+    public static final LatLng INITIAL_CENTER = new LatLng(0, -30);
```

Comment:
Yeah, it wouldn't be a bad idea to remove some of the duplication between the `expandBounds` methods.  Is there other duplication you see that is most need of attention?

Reply:
Those initialization constants that the comment is around are the ones I least like being duplicated. It would be great to guarantee consistency between implementations -- it would be annoying for users to have initial behavior/settings accidentally become different. They're identical between Mapbox and Google implementations.

I think that beyond `expandBounds`, all of the zoom behavior is the same, as are some of the utility methods like conversion between `MapPoint` and `LatLng`. That's more of an aesthetic and mental overhead concern. It's unlikely they're hiding bugs or will change very much.

Another thing that bothers me somewhat is that Mapbox's `LocationEngine` and related classes have the exact same purpose as Collect's `LocationEngineProvider` and related classes, perhaps with subtle implementation differences. It would be preferable for all that to be consistent.

Reply:
Got it.  Yeah, we can definitely move `INITIAL_ZOOM`, `INITIAL_CENTER`, and `POINT_ZOOM` to a common location.  They're identical between all three implementations!  I've opened an issue: #3090 

The conversion methods unfortunately can't be consolidated, as Google's `LatLng` and Mapbox's `LatLng` are different classes even though they have the same name.

I don't see a `LocationEngineProvider` in the `collect` package—did you mean `LocationClient`?  I agree with the desire to harmonize these location frameworks, though it may be tricky.  Should I open an issue for this as well?

Reply:
> They're identical between all three implementations!

Even better, I didn't even check!

> The conversion methods unfortunately can't be consolidated, as Google's LatLng and Mapbox's LatLng are different classes even though they have the same name.

Darn!

> I don't see a LocationEngineProvider in the collect package—did you mean LocationClient?

Yes, indeed, sorry. I think it would feel weird to rely on `com.mapbox.android.core.location` classes in code that isn't Mapbox related. On the other hand, using `org.odk.collect.android.location.client` classes for the Mapbox implementation also seems unfortunate. So... leave it, I guess?